src/buf.rs

```diff
@@ -1,29 +1,139 @@
 use super::Primitive;
+use std::ops::Deref;
 
-/// Additional methods for byte buffers.
-pub trait BufExt {
-    fn push_val<T: Primitive>(&mut self, value: T);
-    fn push_int(&mut self, value: i32);
-    fn push_float(&mut self, value: f32);
-    fn push_decimal(&mut self, value: f32);
-    fn push_hex(&mut self, value: u8);
-    fn push_hex_u16(&mut self, value: u16);
-    fn push_octal(&mut self, value: u8);
+/// Track the limits of data types used in a buffer.
+#[derive(Clone, PartialEq, Debug, Default)]
+pub struct Limits {
+    int: i32,
+    real: f32,
+    name_len: usize,
+    str_len: usize,
+    array_len: usize,
+    dict_entries: usize,
 }
 
-impl BufExt for Vec<u8> {
+impl Limits {
+    /// Create a new `Limits` struct with all values initialized to zero.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub(crate) fn register_int(&mut self, val: i32) {
+        self.int = self.int.max(val.abs());
+    }
+
+    pub(crate) fn register_real(&mut self, val: f32) {
+        self.real = self.real.max(val.abs());
+    }
+
+    pub(crate) fn register_name_len(&mut self, len: usize) {
+        self.name_len = self.name_len.max(len);
+    }
+
+    pub(crate) fn register_str_len(&mut self, len: usize) {
+        self.str_len = self.str_len.max(len);
+    }
+
+    pub(crate) fn register_array_len(&mut self, len: usize) {
+        self.array_len = self.array_len.max(len);
+    }
+
+    pub(crate) fn register_dict_entries(&mut self, len: usize) {
+        self.dict_entries = self.dict_entries.max(len);
+    }
+
+    /// Get the absolute value of the largest positive/negative integer number.
+    pub fn int(&self) -> i32 {
+        self.int
+    }
+
+    /// Get the absolute value of the largest positive/negative real number.
+    pub fn real(&self) -> f32 {
+        self.real
+    }
+
+    /// Get the maximum length of any used name.
+    pub fn name_len(&self) -> usize {
+        self.name_len
+    }
+
+    /// Get the maximum length of any used array.
+    pub fn array_len(&self) -> usize {
+        self.array_len
+    }
+
+    /// Get the maximum number of entries in any dictionary.
+    pub fn dict_entries(&self) -> usize {
+        self.dict_entries
+    }
+
+    /// Get the maximum length of any used string.
+    pub fn str_len(&self) -> usize {
+        self.str_len
+    }
+
+    /// Merge two `Limits` with each other, taking the maximum
+    /// of each field from both.
+    pub fn merge(&mut self, other: &Limits) {
+        self.register_int(other.int);
+        self.register_real(other.real);
+        self.register_name_len(other.name_len);
+        self.register_str_len(other.str_len);
+        self.register_array_len(other.array_len);
+        self.register_dict_entries(other.dict_entries);
+    }
+}
+
+/// A buffer of arbitrary PDF content.
+#[derive(Clone, PartialEq, Debug)]
+pub struct Buf {
+    pub(crate) inner: Vec<u8>,
+    pub(crate) limits: Limits,
+}
+
+impl Deref for Buf {
+    type Target = Vec<u8>;
```

Comment:
Should probably rather deref to `[u8]`. Could you also move the impl below the inherent impl?